### PR TITLE
Implement 26 EXPERT1 cards

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Implement 26 cards in the EXPERT1 card set
    - Siegebreaker (EX1_185)
    - SI:7 Infiltrator (EX1_186)
    - Arcane Devourer (EX1_187)
    - Barrens Stablehand (EX1_188)
    - Brightwing (EX1_189)
    - High Inquisitor Whitemane (EX1_190)
    - Lightning Storm (EX1_259)
    - Cone of Cold (EX1_275)
    - Ice Barrier (EX1_289)
    - Mirror Entity (EX1_294)
    - Void Terror (EX1_304)
    - Summoning Portal (EX1_315)
    - Lord Jaraxxus (EX1_323)
    - Shadow Madness (EX1_334)
    - Lightspawn (EX1_335)
    - Thoughtsteal (EX1_339)
    - Prophet Velen (EX1_350)
    - Blessing of Wisdom (EX1_363)
    - Sword of Justice (EX1_366)
    - Repentance (EX1_379)
    - Arathi Weaponsmith (EX1_398)
    - Armorsmith (EX1_402)
    - Upgrade! (EX1_409)
    - Gorehowl (EX1_411)
    - Scavenging Hyena (EX1_531)
    - Misdirection (EX1_533)
- Create class `OngoingEnchant`: This class is implementation of a kind of enchantment that its effect gradually grows due to a trigger. OngoingEnchant is narrowly used when the source of the trigger and the target of the Enchantment is identical. (e.g. Mana Wyrm)
- Create class `SummoningPortalAura`: This class is an aura for card 'Summoning Portal'
- Create class `PlayerAuraEffects`: This class is a collection of controller Tag increments from Auras. Note that these tags tends to be checked when a player plays any cards
- Create several task classes
    - Class `CustomTask`: This class represents the task for running custom function
    - Class `DamageWeaponTask`: This class represents the task for taking damage of durability to one of the equipped weapons
    - Class `RandomMinionTask`: This class represents the task for picking minion(s) at random
    - Class `ReplaceHeroTask`: This class represents the task for replacing the hero
    - Class `SummonCopyTask`: This class represents the task for summoning a copy of one (or more) existing entity
- Refactor several tasks
    - Class `DamageTask`: Add variable `randomDamage` and code to process it
    - Class `FilterStackTask`: Change code to take multiple self or relation conditions
    - Class `ConditionTask`: Change type to `SelfCondition*` or `RelaCondition*`
- Add some methods
    - Method `IsFrozen()`: Returns the flag that indicates whether it is frozen
    - Method `IsEventTargetIs()`: SelfCondition wrapper for checking the event target is cardType
    - Method `IsProposedDefender()`: SelfCondition wrapper for checking it is proposed defender
    - Method `IsNotUntouchable()`: SelfCondition wrapper for checking the entity is not untouchable
    - Method `CopyInternalAttributes()`: Copies internal attributes to copy
    - Method `IsOpFieldNotFull()`: SelfCondition wrapper for checking the opponent field is not full
    - Method `GetCount()`: Gets the count of the elements
    - Method `GetCost()`: Returns the value of cost
    - Method `QueueUpdate()`: Queues the update
    - Method `GetType()`: Gets the type of aura
    - Method `HasCharget()`: Returns the flag that indicates whether it has charge
    - Method `IsNotImmune()`: SelfCondition wrapper for checking the entity is not immune
    - Method `IsInZone()`: SelfCondition wrapper for checking the entity is in zone
- Add several types
    - Aura type `PLAYER`
    - Entity type `ENEMY_SECRETS`
- Create file `GameTagCustom.def` to separate custom game tag

### CODE REVIEW

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/utilForever/RosettaStone/pullRequest?prid=4635514&bid=15533076)

File: Sources/Rosetta/Actions/Summon.cpp [hunk 1]
```diff
@@ -9,18 +9,27 @@
...
+    game->taskQueue.StartEvent();
+    EventMetaData* temp = game->currentEventData;
+    if (summoner != nullptr)
+    {
+        game->currentEventData =
+            new EventMetaData(dynamic_cast<Playable*>(summoner), minion);
+    }
+    game->triggerManager.OnAfterSummonTrigger(minion);
+    game->ProcessTasks();
+    game->currentEventData = temp;
```
Comment (utilForever) [inline Sources/Rosetta/Actions/Summon.cpp]: Codacy Issue found: [Memory leak: currentEventData](https://app.codacy.com/app/utilForever/RosettaStone/pullRequest?prid=4635514)

File: Includes/Rosetta/Tasks/SimpleTasks/ReplaceHeroTask.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+//!
+class ReplaceHeroTask : public ITask
+{
+ public:
+    //! Constructs task with given \p heroCard, \p heroPowerCard
+    //! and \p weaponCard.
+    //! \param heroCard The hero card to replace.
+    //! \param heroPowerCard The hero power card to replace.
+    //! \param weaponCard The weapon card to equip.
+    explicit ReplaceHeroTask(const std::string& heroCard,
```
Comment (CreatorSeraph) [inline Includes/Rosetta/Tasks/SimpleTasks/ReplaceHeroTask.hpp]: string_view의 사용을 고려해볼 수 있을까요?

File: Sources/Rosetta/Conditions/SelfCondition.cpp [hunk 3]
```diff
@@ -39,13 +39,40 @@ SelfCondition SelfCondition::IsNotDead()
         [=](Playable* playable) -> bool { return !playable->isDestroyed; });
 }
 
+SelfCondition SelfCondition::IsNotImmune()
+{
+    return SelfCondition([=](Playable* playable) -> bool {
```
Comment (CreatorSeraph) [inline Sources/Rosetta/Conditions/SelfCondition.cpp]: 람다 캡쳐부분에 =를 넣은 이유가 있나요?

File: Sources/Rosetta/Tasks/SimpleTasks/MathNumberIndexTask.cpp [hunk 4]
```diff
@@ -52,6 +52,15 @@ TaskStatus MathNumberIndexTask::Impl(Player* player)
...
             stack.num1 = result;
             break;
+        case 2:
+            stack.num2 = result;
+            break;
+        case 3:
+            stack.num3 = result;
+            break;
+        case 4:
+            stack.num4 = result;
```
Comment (CreatorSeraph) [inline Sources/Rosetta/Tasks/SimpleTasks/MathNumberIndexTask.cpp]: num들을 array를 사용하는것도 괜찮을것같습니다

File: Includes/Rosetta/Tasks/SimpleTasks/ReplaceHeroTask.hpp (hunk 2 above)
Comment (utilForever) [reply Includes/Rosetta/Tasks/SimpleTasks/ReplaceHeroTask.hpp]: 좋은 피드백입니다. 해당 클래스의 생성자 뿐만 아니라 `FindCardByID`와 `FindCardByName` 함수의 인자 타입도 `std::string_view`로 수정했습니다.

File: Sources/Rosetta/Conditions/SelfCondition.cpp (hunk 3 above)
Comment (utilForever) [reply Sources/Rosetta/Conditions/SelfCondition.cpp]: 다른 함수도 전부 `=`로 캡쳐했습니다. 그 이유는 `&`로 하면 원하는 대로 카드 로직을 구현할 수 없기 때문입니다.

File: Sources/Rosetta/Tasks/SimpleTasks/MathNumberIndexTask.cpp (hunk 4 above)
Comment (utilForever) [reply Sources/Rosetta/Tasks/SimpleTasks/MathNumberIndexTask.cpp]: `std::array` 타입으로 변경했습니다.

**Pull request merged**